jlm/hls/backend/rvsdg2rhls/stream-conv.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2025 David Metz <david.c.metz@ntnu.no>
+ * See COPYING for terms of redistribution.
+ */
+
+#include <jlm/hls/backend/rvsdg2rhls/hls-function-util.hpp>
+#include <jlm/hls/backend/rvsdg2rhls/rhls-dne.hpp>
+#include <jlm/hls/backend/rvsdg2rhls/stream-conv.hpp>
+#include <jlm/hls/ir/hls.hpp>
+#include <jlm/rvsdg/lambda.hpp>
+
+#include <algorithm>
+
+namespace jlm::hls
+{
+
+void
+stream_conv(llvm::RvsdgModule & rm)
+{
+  auto & graph = rm.Rvsdg();
+  auto root = &graph.GetRootRegion();
+  auto lambda = dynamic_cast<rvsdg::LambdaNode *>(root->Nodes().begin().ptr());
+  auto stream_enqs = find_function_arguments(lambda, "hls_stream_enq");
+  auto stream_deqs = find_function_arguments(lambda, "hls_stream_deq");
+  if (stream_enqs.empty())
+  {
+    JLM_ASSERT(stream_deqs.empty());
+    return;
+  }
+  std::vector<rvsdg::SimpleNode *> enq_calls, deq_calls;
+  std::unordered_set<rvsdg::output *> visited;
+  for (auto stream_enq : stream_enqs)
+  {
+    JLM_ASSERT(stream_enq.inner);
+    trace_function_calls(stream_enq.inner, enq_calls, visited);
+    visited.erase(visited.begin(), visited.end());
+  }
+  for (auto stream_deq : stream_deqs)
+  {
+    trace_function_calls(stream_deq.inner, deq_calls, visited);
+    visited.erase(visited.begin(), visited.end());
+  }
+  JLM_ASSERT(!enq_calls.empty());
+  JLM_ASSERT(!deq_calls.empty());
+  for (auto & enq_call : enq_calls)
+  {
+    auto enq_constant = trace_constant(enq_call->input(1)->origin());
+    for (auto & deq_call : deq_calls)
+    {
+      auto deq_constant = trace_constant(deq_call->input(1)->origin());
+      if (*enq_constant == *deq_constant)
```

Comment:
This if is doing a lot of heavy lifting, could everything inside it be moved to a separate function that handles a specific (enq, deq) pair?

Reply:
Moved to separate function